### PR TITLE
Admin tool to remove a package from its publisher

### PR DESCRIPTION
**Pull request by sigurdm**
The package gets the publishers owners as uploaders

https://github.com/dart-lang/pub-dev/issues/6357

### CODE REVIEW

File: app/lib/admin/tools/remove_package_from_publisher.dart [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import 'package:pub_dev/package/models.dart';
+import 'package:pub_dev/publisher/backend.dart';
+import 'package:pub_dev/shared/datastore.dart';
+
+final _argParser = ArgParser()
+  ..addOption('package', help: 'The package to update.')
+  ..addOption('publisher-id', help: 'The `publisherId` to set.')
+  ..addFlag('help', abbr: 'h', defaultsTo: false, help: 'Show help.');
+
+Future<String> executeRemovePackageFromPublisher(List<String> args) async {
```
Comment (jonasfj) [inline app/lib/admin/tools/remove_package_from_publisher.dart:18]: Would be nicer to make admin actions:
https://github.com/dart-lang/pub-dev/blob/master/app/lib/admin/actions/actions.dart

We don't have CLI for admin-actions yet, but it should be easy to build.

**Review by jonasfj — APPROVED** [review]

File: app/lib/admin/tools/remove_package_from_publisher.dart (hunk 1 above)
Comment (sigurdm) [reply app/lib/admin/tools/remove_package_from_publisher.dart]: Ah yes, I forgot you landed these. Much nicer!

Done

File: app/lib/admin/tools/remove_package_from_publisher.dart [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  final currentPublisherMembers =
+      (await publisherBackend.listPublisherMembers(currentPublisherId));
+
+  await withRetryTransaction(dbService, (tx) async {
+    final pkg = await tx.lookupValue<Package>(package.key);
+    pkg.publisherId = null;
+    pkg.uploaders = currentPublisherMembers.map((e) => e.userId).toList();
+    pkg.updated = clock.now().toUtc();
+    tx.insert(pkg);
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply app/lib/admin/tools/remove_package_from_publisher.dart]: Should we instead keep a separate log of all admin-actions and their output?

**Review by isoos — APPROVED** [review]

File: app/lib/admin/tools/remove_package_from_publisher.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonasfj) [reply app/lib/admin/tools/remove_package_from_publisher.dart]: > Hmm, what should it say?

The action taken and that `support@pub.dev` did it.

File: app/lib/admin/actions/remove_package_from_publisher.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    options: {
+      'package': 'name of the package to remove from its current publisher'
+    },
+    summary:
+        'Removes <package> from its current publisher, and makes that publisher\'s owners uploaders',
+    description: '''
+Removes <package> from its current publisher, and makes that publisher\'s owners
+ uploaders.
+
+Fails if that package currently has no uploader.
```
Comment (jonasfj) [inline app/lib/admin/actions/remove_package_from_publisher.dart:24]: This doesn't make sense, should it be negated?

File: app/lib/admin/actions/remove_package_from_publisher.dart [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+Example: given a publisher example.com with two members a@example.com and
+ b@example.com, and apackage:foo in the publisher.
```
Comment (jonasfj) [inline app/lib/admin/actions/remove_package_from_publisher.dart:27]:
```suggestion
 b@example.com, and a package:foo in the publisher.
```

**Review by jonasfj — APPROVED** [review]

File: app/lib/admin/tools/remove_package_from_publisher.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (sigurdm) [reply app/lib/admin/tools/remove_package_from_publisher.dart]: Done

File: app/lib/admin/actions/remove_package_from_publisher.dart (hunk 3 above)
Comment (sigurdm) [reply app/lib/admin/actions/remove_package_from_publisher.dart]: Good catch, should be publisher

**Pull request merged**